presto-main/src/main/java/com/facebook/presto/execution/buffer/SpoolingOutputBuffer.java

```diff
@@ -0,0 +1,747 @@
+
+    @Override
+    public OutputBufferInfo getInfo()
+    {
+        return new OutputBufferInfo(
+                "SPOOLING",
+                state.get(),
+                state.get().canAddBuffers(),
+                state.get().canAddPages(),
+                totalBufferedBytes.get(),
+                totalBufferedPages.get(),
+                totalRowsAdded.get(),
+                totalPagesAdded.get(),
+                ImmutableList.of()
+        );
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return state.get() == BufferState.FINISHED;
+    }
+
+    @Override
+    public double getUtilization()
+    {
+        return 0;
+    }
+
+    @Override
+    public boolean isOverutilized()
+    {
+        return false;
+    }
+
+    @Override
+    public long getPeakMemoryUsage()
+    {
+        return 0;
+    }
+
+    @Override
+    public ListenableFuture<?> isFull()
+    {
+        return immediateFuture(null);
+    }
+
+    @Override
+    public void addStateChangeListener(StateMachine.StateChangeListener<BufferState> stateChangeListener)
+    {
+        state.addStateChangeListener(stateChangeListener);
+    }
+
+    @Override
+    public void setOutputBuffers(OutputBuffers newOutputBuffers)
+    {
+        requireNonNull(newOutputBuffers, "newOutputBuffers is null");
+        checkArgument(outputBuffers.getType() == OutputBuffers.BufferType.SPOOLING, "Invalid output buffers type");
+        checkArgument(outputBuffers.isNoMoreBufferIds(), "invalid noMoreBufferIds");
+
+        if (state.get().isTerminal() || outputBuffers.getVersion() >= newOutputBuffers.getVersion()) {
+            return;
+        }
+
+        outputBuffers.checkValidTransition(newOutputBuffers);
+    }
+
+    @Override
+    public void enqueue(Lifespan lifespan, List<SerializedPage> pages)
+    {
+        PendingRead pendingRead;
+        synchronized (this) {
+            if (!state.get().canAddPages()) {
+                return;
+            }
+
+            this.pages.addAll(pages);
+
+            // update output buffer info
+            totalBufferedBytes.addAndGet(getPagesSize(pages));
+            totalBufferedPages.addAndGet(pages.size());
+            totalPagesAdded.addAndGet(pages.size());
+            totalRowsAdded.addAndGet(getPagesRows(pages));
+
+            totalPagesRemaining.addAndGet(pages.size());
+
+            if (getPagesSize(this.pages) >= threshold.toBytes()) {
+                flush();
+            }
+
+            pendingRead = this.pendingRead;
+            this.pendingRead = null;
+        }
+
+        if (pendingRead != null) {
+            processPendingRead(pendingRead);
+        }
+    }
+
+    @Override
+    public synchronized void enqueue(Lifespan lifespan, int partition, List<SerializedPage> pages)
+    {
+        checkState(partition == 0, "Expected partition number to be zero");
+        enqueue(lifespan, pages);
+    }
+
+    private synchronized void flush()
+    {
+        if (pages.isEmpty()) {
```

Comment:
I don't think we need that as it is synchronized private method only used by enqueue. And it checks the threshold before calling this method, am I wrong?